### PR TITLE
Add unit tests for operator_installer.go

### PR DESCRIPTION
**Pull request by asmacdo**
First pass, adding unit tests to operator_installer.go

- [ ] Remove TODO items before merge

### CODE REVIEW

**Comment by asmacdo** [comment]
/hold 
until TODO questions are removed

File: go.mod
Comment (jmrodri) [inline go.mod:7]: Remove `github.com/fatih/color`

File: internal/olm/operator/registry/olm_resources.go
Comment (jmrodri) [inline internal/olm/operator/registry/olm_resources.go:74]: It is quite weird. Basically it's a functional option pattern. https://www.sohamkamani.com/golang/options-pattern/ The idea is that you can supply optional items when constructing an instance. Often these options are of very different types, as in the example of the `House` in the url above they wanted to pass in `WithConcrete` which is a `string` and `WithFireplace` which is a `bool`.

WIthout this pattern, you would either have to have a lengthy constructor function that you always have to pass in every option (even if they are optional). i.e. `NewHouse(material string, hasfireplace bool, floors int, foo Foo, bar *Bar, etc)`. It also means every time you want to extend it you'd likely have to add more to the constructor function.

With the options pattern the option is a function. In our case for subscription we did not define a new Option type. If we did it would look like this: `type SubscriptionOption func(*v1alpha1.Subscription)`. So we're just returning a function that takes in a `*v1alpha1.Subscription`. 

So with this pattern, we can create `WithSomeOption` functions that take in the specific type we want and by returning a function it can be easily used by the constructor function. So in the above code you see we have the following:

```golang
func withInstallPlanApproval(approval v1alpha1.Approval) func(*v1alpha1.Subscription) {
	return func(sub *v1alpha1.Subscription) {
		if sub.Spec == nil {
			sub.Spec = &v1alpha1.SubscriptionSpec{}
		}
		sub.Spec.InstallPlanApproval = approval
	}
}
```
That will take in an Approval which is an attribute we want to set on the `Subscription`. So we return a function that expects a `Subscription` to be passed in so we can modify it. If I want to override the `PackageName` I could easily create the following to update `PackageName`:

```golang
func withPackageName(pkgName string) func(*v1alpha1.Subscription) {
	return func(sub *v1alpha1.Subscription) {
		if sub.Spec == nil {
			sub.Spec = &v1alpha1.SubscriptionSpec{}
		}
		sub.Spec.PackageName = pkgName
	}
}
```

I don't have to change the constructor function since it takes in a variable list of "options" which is a list of `func(*v1alpha1.Subscription)`.

File: internal/olm/operator/registry/operator_installer.go
Comment (jmrodri) [inline internal/olm/operator/registry/operator_installer.go:216]: Exactly what I described above regarding Options pattern. So now you can clearly see that we are creating a `Subscription` instance and setting it's `PackageName`, `CatalogSource`, and `InstallPlan`. And if we want to update a fourth attribute, we can just add a new `withThatNewAttribute` and not have to change the `newSubscription`.

File: internal/olm/operator/registry/operator_installer.go
Comment (jmrodri) [inline internal/olm/operator/registry/operator_installer.go:212]: +1 to not passing in an entire struct just to get the name.

File: internal/olm/operator/registry/operator_installer_test.go
Comment (jmrodri) [inline internal/olm/operator/registry/operator_installer_test.go:77]: this is weird. We shouldn't have to use a real kubeconfig since we're using a fake client. Should be able to do the following:

```golang
cfg := &operator.Configuration{}
...
cfg.Client = fake.NewFakeClientWithScheme(sch)
```

File: internal/olm/operator/registry/operator_installer_test.go
Comment (jmrodri) [inline internal/olm/operator/registry/operator_installer_test.go:124]: Not in this PR. I think we need to refactor the code to make it testable. I don't like when we create clients there:
```
c, err := olmclient.NewClientForConfig(o.cfg.RESTConfig)
```
We can do the option pattern in this method:

```golang
type Option func(...)

func WithOLMClient(...)  {
}

func (o OperatorInstaller) getInstalledCSV(ctx context.Context, opts...lOption) (*v1alpha1.ClusterServiceVersion, error) {
...
}
```

Then in the test we can create a fake client and pass it in via the `WithOLMClient`. In the main line code, it can create the client like it does. We had to do something similar to the leader election code in operator-lib:
* https://github.com/operator-framework/operator-lib/blob/main/leader/leader.go#L91-L104
* https://github.com/operator-framework/operator-lib/blob/main/leader/leader.go#L53-L82

File: internal/olm/operator/registry/operator_installer_test.go
Comment (jmrodri) [inline internal/olm/operator/registry/operator_installer_test.go:132]: See previous comment about this

File: internal/olm/operator/registry/operator_installer_test.go
Comment (jmrodri) [inline internal/olm/operator/registry/operator_installer_test.go:145]: since you are not using existingIp anywhere else you can initialize the fake client inline:
```golang

		oi.cfg.Client = fake.NewFakeClientWithScheme(sch, 
			&v1alpha1.InstallPlan{
				ObjectMeta: metav1.ObjectMeta{
					Name:      "fakeName",
					Namespace: "fakeNS",
				},
			},
		)
```

File: internal/olm/operator/registry/operator_installer_test.go
Comment (jmrodri) [inline internal/olm/operator/registry/operator_installer_test.go:158]: I don't think you need to look up the install plan you just created, at least here. I would do this *after* calling `approveInstallPlan`.

File: internal/olm/operator/registry/operator_installer_test.go
Comment (jmrodri) [inline internal/olm/operator/registry/operator_installer_test.go:194]: You could test one scenario by having an EMPTY fake client:

```golang
    if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
        if err := o.cfg.Client.Get(ctx, ipKey, &ip); err != nil {
            return fmt.Errorf("error getting install plan: %v", err)
        }
        ...
```
If you use an empty fakeclient the get will fail with `NotFound` error. Then you can validate that the error contains "error getting install plan".

The `Update` part of the code is harder, not sure yet the best way to do that.

File: internal/olm/operator/registry/operator_installer_test.go
Comment (jmrodri) [inline internal/olm/operator/registry/operator_installer_test.go:205]: no need for kubeconfigPath

File: internal/olm/operator/registry/operator_installer_test.go
Comment (jmrodri) [inline internal/olm/operator/registry/operator_installer_test.go:244]: I would probably put a blank line before the `waitForInstallPlan` so that it is easier to see the stuff above it are for setup.

File: internal/olm/operator/registry/olm_resources.go
Comment (asmacdo) [reply internal/olm/operator/registry/olm_resources.go:74]: Thats pretty slick! Thanks!

File: internal/olm/operator/registry/operator_installer_test.go
Comment (asmacdo) [reply internal/olm/operator/registry/operator_installer_test.go:158]: Thats was just to prove to myself that I wasn't getting a false pass. (ie, somehow the Spec was approved without calling the function). I don't mind removing it.

File: internal/olm/operator/registry/operator_installer.go
Comment (asmacdo) [reply internal/olm/operator/registry/operator_installer.go:216]: This makes SO much more sense now.

File: internal/olm/operator/registry/operator_installer_test.go
Comment (asmacdo) [reply internal/olm/operator/registry/operator_installer_test.go:124]: @jmrodri I'll file an issue.

File: go.mod
Comment (camilamacedo86) [reply go.mod:7]: it still required to get done.

File: internal/olm/operator/registry/operator_installer.go
Comment (camilamacedo86) [inline internal/olm/operator/registry/operator_installer.go:22]: remove ^

File: internal/olm/operator/registry/operator_installer.go
Comment (camilamacedo86) [inline internal/olm/operator/registry/operator_installer.go:222]: Is this comment required? WhY?

File: internal/olm/operator/registry/operator_installer_test.go
Comment (camilamacedo86) [inline internal/olm/operator/registry/operator_installer_test.go:34]: remove the comments

File: internal/olm/operator/registry/operator_installer_test.go
Comment (camilamacedo86) [inline internal/olm/operator/registry/operator_installer_test.go:46]: missing comment because it is an exported func. Also, why it needs to be exported?

Comment (asmacdo) [reply internal/olm/operator/registry/operator_installer_test.go:46]: It doesnt. nice catch

File: internal/olm/operator/registry/operator_installer_test.go
Comment (asmacdo) [reply internal/olm/operator/registry/operator_installer_test.go:194]: I tested the empty client 1 test up. (Do we prefer HappyPath tests before err tests?)

File: internal/olm/operator/registry/operator_installer_test.go
Comment (asmacdo) [reply internal/olm/operator/registry/operator_installer_test.go]: https://github.com/operator-framework/operator-sdk/issues/4173

File: internal/olm/operator/registry/operator_installer.go
Comment (asmacdo) [reply internal/olm/operator/registry/operator_installer.go]: Its an artifact for development, I left it in just to help me test, but always planned to remove before merge.  Its gone now.

**Review by jmrodri — APPROVED** [review]
/lgtm

**Review by camilamacedo86 — APPROVED** [review]
Please, just not forget to remove the todo from the first commit. Otherwise 🥇 

/lgtm
/approve

**Pull request merged**
